prometheus/internalmetrics/internal_metrics.go

```diff
@@ -0,0 +1,202 @@
+// Package internalmetrics provides functionality to collect Prometheus metrics.
+package internalmetrics
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	// Because this package is used all throughout the codebase, be VERY careful adding new
+	// kiali imports here. Most likely you will encounter an import cycle error that will
+	// cause a compilation failure.
+)
+
+// These constants define the different label names for the different metric timeseries
+const (
+	labelGraphKind        = "graph_kind"
+	labelGraphType        = "graph_type"
+	labelWithServiceNodes = "with_service_nodes"
+	labelAppender         = "appender"
+	labelName             = "name"
+	labelQueryID          = "query_id"
+)
+
+// MetricsType defines all of Kiali's own internal metrics.
+type MetricsType struct {
+	GraphsGenerated          *prometheus.CounterVec
+	GraphNodes               *prometheus.GaugeVec
+	GraphGenerationTime      *prometheus.HistogramVec
+	GraphAppenderTime        *prometheus.HistogramVec
+	GraphMarshalTime         *prometheus.HistogramVec
+	APIProcessingTime        *prometheus.HistogramVec
+	PrometheusProcessingTime *prometheus.HistogramVec
+}
+
+// Metrics contains all of Kiali's own internal metrics.
+// These metrics can be accessed directly to update their values, or
+// you can use available utility functions defined below.
+var Metrics = MetricsType{
+	GraphsGenerated: prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "kiali_graphs_generated_total",
+			Help: "The total number of graphs Kiali has generated.",
+		},
+		[]string{labelGraphKind, labelGraphType, labelWithServiceNodes},
+	),
+	GraphNodes: prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "kiali_graph_nodes",
+			Help: "The number of nodes in a generated graph.",
+		},
+		[]string{labelGraphKind, labelGraphType, labelWithServiceNodes},
+	),
+	GraphGenerationTime: prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "kiali_graph_generation_duration_seconds",
+			Help: "The time required to generate a graph.",
+		},
+		[]string{labelGraphKind, labelGraphType, labelWithServiceNodes},
+	),
+	GraphAppenderTime: prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "kiali_graph_appender_duration_seconds",
+			Help: "The time required to execute an appender while generating a graph.",
+		},
+		[]string{labelAppender},
+	),
+	GraphMarshalTime: prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "kiali_graph_marshal_duration_seconds",
+			Help: "The time required to marshal and return the JSON for a graph.",
+		},
+		[]string{labelGraphKind, labelGraphType, labelWithServiceNodes},
+	),
+	APIProcessingTime: prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name: "kiali_api_processing_duration_seconds",
```

Comment:
could we rename "api" => "rest_api" ? I didn't understand exactly what it was before seeing in code. (API could also be calls to underlying apis, such as k8s')

Reply:
I can do that. It does seem it could cause confusion; however, to be clear, the reason why I called it "api" is because it represents the rest calls going to the "/api" URL endpoints.

Reply:
From my outsider POV, `kiali_api_processing_duration_seconds` looks fine. If it were for measuring timing to downstream services, I would expect something like `kiali_kuberbenetes_api_processing_duration_seconds` (eg `<app>_<subsystem>_<metric>`).